bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/services/seo/LanguageNavigationSiteRootSelectionStrategy.java

```diff
@@ -0,0 +1,140 @@
+ ~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ ~ See the License for the specific language governing permissions and
+ ~ limitations under the License.
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/
+package com.adobe.cq.wcm.core.components.internal.services.seo;
+
+import java.util.Optional;
+import java.util.concurrent.ExecutionException;
+import java.util.function.Predicate;
+
+import org.apache.sling.api.resource.Resource;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import org.osgi.framework.Constants;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.ConfigurationPolicy;
+import org.osgi.service.component.annotations.Reference;
+import org.osgi.service.component.annotations.ReferenceCardinality;
+import org.osgi.service.component.annotations.ReferencePolicyOption;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.adobe.aem.wcm.seo.localization.SiteRootSelectionStrategy;
+import com.adobe.cq.wcm.core.components.internal.Utils;
+import com.adobe.cq.wcm.core.components.internal.models.v1.ExperienceFragmentDataImpl;
+import com.adobe.cq.wcm.core.components.models.LanguageNavigation;
+import com.day.cq.wcm.api.LanguageManager;
+import com.day.cq.wcm.api.Page;
+import com.day.cq.wcm.api.TemplatedResource;
+import com.day.cq.wcm.msm.api.LiveRelationshipManager;
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+
+/**
+ * An implementation of {@link SiteRootSelectionStrategy} that looks for a language navigation component on the given page and uses it's
+ * configured navigation root as site root.
+ * <p>
+ * This Component must be explicitly enabled by providing an (empty) configuration for it. This should be the case for sites the use the
+ * language navigation core component. For any other case this Component should be kept disabled.
+ */
+@Component(
+    configurationPolicy = ConfigurationPolicy.REQUIRE,
+    property = {
+        Constants.SERVICE_RANKING + ":Integer=100"
+    },
+    service = { SiteRootSelectionStrategy.class }
+)
+public class LanguageNavigationSiteRootSelectionStrategy implements SiteRootSelectionStrategy {
+
+    private static final Logger LOG = LoggerFactory.getLogger(LanguageNavigationSiteRootSelectionStrategy.class);
+
+    @Reference(
+        cardinality = ReferenceCardinality.OPTIONAL,
+        policyOption = ReferencePolicyOption.GREEDY,
+        target = "(objectClass=*DefaultSiteRootSelectionStrategy)")
+    private SiteRootSelectionStrategy defaultSelectionStrategy;
+
+    @Reference
+    private LanguageManager languageManager;
+
+    @Reference
+    private LiveRelationshipManager liveRelationshipManager;
+
+    private final Cache<Page, Optional<Resource>> languageNavigationCache = CacheBuilder.newBuilder().weakKeys().build();
```

Comment:
A cache with the page key should not be used in combination with weakKeys as this will end up in a `==` comparisons and will create new keys without needed (https://guava.dev/releases/18.0/api/docs/com/google/common/cache/CacheBuilder.html)

Reply:
This is by intent to keep the complexity small (no need for an eviction strategy, cache entry live-time or max-size etc.). It should only address cases where the service is called with the same page object multiple times, but was not supposed to introduce a long-living caching beyond the scope of a request. 

I would recommend to keep it like that as this addresses a proven performance concern and improve it when we see an actual need.